### PR TITLE
Add stopwatches

### PR DESCRIPTION
**Pull request by anastasds**
Adds the ability to measure execution time of arbitrary code blocks. `CuspStopwatch` is implemented as `Observable` and `AutoCloseable` with `CuspObserver` automatically notified at the end of a try-with-resources block containing code to be timed. `CuspObserver` is abstract to allow the library user todo what they like with the measurement.

Complements the `StageOutcomeListener`, which is for reporting the execution times of entire stages.

We can add more general reporting mechanisms using something like an abstract `CuspEvent` when we have a need for it.

### CODE REVIEW

File: src/main/java/com/ibm/cusp/graph/observe/CuspStopwatch.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    public Duration getDuration() {
+        return duration;
+    }
+
+    @Override
+    public void close() {
+        duration = Duration.ofMillis(System.currentTimeMillis() - startTimeMillis);
+
+        if(countObservers() > 0) {
+            logger.trace("Stopped stopwatch {} at {}ns after {}ns", identifier, startTimeMillis, duration.toMillis());
```
Comment (jaymindesai) [inline src/main/java/com/ibm/cusp/graph/observe/CuspStopwatch.java]: This should be `ms`?

File: src/main/java/com/ibm/cusp/graph/observe/CuspStopwatch.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+            setChanged();
+
+            logger.trace("Notifying observer");
+            notifyObservers();
+        } else {
+            logger.warn("Stopwatch completed but there are no observers to report to");
+            logger.trace("Stopwatch {} stopped after {}ns", identifier, duration.toMillis());
```
Comment (jaymindesai) [inline src/main/java/com/ibm/cusp/graph/observe/CuspStopwatch.java]: This should be `ms`?

File: src/main/java/com/ibm/cusp/graph/observe/CuspStopwatch.java (hunk 1 above)
Comment (jaymindesai) [reply src/main/java/com/ibm/cusp/graph/observe/CuspStopwatch.java]: Shouldn't the second placeholder in this log be the time the stopwatch was actually stopped?

Comment (anastasds) [reply src/main/java/com/ibm/cusp/graph/observe/CuspStopwatch.java]: Yes, fixed and fixed.

File: src/main/java/com/ibm/cusp/graph/observe/CuspStopwatch.java (hunk 2 above)
Comment (anastasds) [reply src/main/java/com/ibm/cusp/graph/observe/CuspStopwatch.java]: Fixed.

File: src/main/java/com/ibm/cusp/graph/observe/CuspStopwatch.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+     */
+    public CuspStopwatch(CuspObserver observer, String stopwatchIdentifier) {
+        identifier = stopwatchIdentifier;
+        startTimeMillis = System.currentTimeMillis();
+
+        if(observer != null) {
+            addObserver(observer);
+        }
+
+        logger.trace("Started stopwatch {} at {}ns", stopwatchIdentifier, startTimeMillis);
```
Comment (jaymindesai) [inline src/main/java/com/ibm/cusp/graph/observe/CuspStopwatch.java]: One last `ns` -> `ms`

**Comment by anastasds** [comment]
Tests pass locally; next up will be adding a build check.

**Review by jaymindesai — APPROVED** [review]
Looks good

**Pull request merged**